crypto/crmf/crmf_lib.c

```diff
@@ -673,32 +675,36 @@ X509 *OSSL_CRMF_ENCRYPTEDVALUE_get1_encCert(OSSL_CRMF_ENCRYPTEDVALUE *ecert,
                 CRMF_R_UNSUPPORTED_CIPHER);
         return NULL;
     }
-
+    /* select symmetric cipher based on algorithm given in message */
+    if ((cipher = EVP_get_cipherbynid(symmAlg)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_ENCRYPTEDVALUE_GET1_ENCCERT,
+                CRMF_R_UNSUPPORTED_CIPHER);
+        goto end;
+    }
+    ERR_clear_error();
+    cikeysize = EVP_CIPHER_key_length(cipher);
     /* first the symmetric key needs to be decrypted */
     pkctx = EVP_PKEY_CTX_new(pkey, NULL);
     if (pkctx != NULL && EVP_PKEY_decrypt_init(pkctx)) {
         ASN1_BIT_STRING *encKey = ecert->encSymmKey;
-        size_t eksize = 0;
-
-        if (EVP_PKEY_decrypt(pkctx, NULL, &eksize, encKey->data, encKey->length)
-                <= 0
-                || (ek = OPENSSL_malloc(eksize)) == NULL
-                || EVP_PKEY_decrypt(pkctx, ek, &eksize, encKey->data,
-                                    encKey->length) <= 0) {
+        int decrypt_failed = EVP_PKEY_decrypt(pkctx, NULL, &eksize,
+                                              encKey->data,
+                                              encKey->length) <= 0;
+        if ((ek = OPENSSL_malloc(eksize)) == NULL) {
+            goto oom;
+        }
+        decrypt_failed |= EVP_PKEY_decrypt(pkctx, ek, &eksize,
+                                           encKey->data, encKey->length) <= 0;
+        decrypt_failed |= eksize != (size_t)cikeysize;
```

Comment:
even better would be using our constant time API here,
```C
int retval = EVP_PKEY_decrypt(pkctx, ek, &eksize,
                                           encKey->data, encKey->length);
ERR_clear_error(); /* error state has sensitive information */
size_t decrypt_failed = constant_time_is_zero_s( constant_time_msb(-retval));
decrypt_failed |= ~constant_time_eq_s (eksize, (size_t)cikeysize));
if (decrypt_failed) {
```

Reply:
this leaves the undefined overflow on negation of int, thus -0x80000000 -> 0x80000000,
maybe we don't care, or we could do something like
`~constant_time_is_zero_s(constant_time_msb(retval) | constant_time_is_zero(retval))`